src/coprocessor/dag/expr/builtin_encryption.rs

```diff
@@ -91,13 +94,95 @@ impl ScalarFunc {
         };
         Ok(Some(Cow::Owned(sha2)))
     }
+
+    #[inline]
+    pub fn compress<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        use byteorder::{ByteOrder, LittleEndian};
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+
+        // according to MySQL doc: Empty strings are stored as empty strings.
+        if input.is_empty() {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        let mut e = ZlibEncoder::new(input.as_ref(), Compression::default());
+        // prefered capacity is input length plus four bytes length header and one extra end "."
+        // max capacity is isize::max_value(), or will panic with "capacity overflow"
+        let mut vec = Vec::with_capacity((input.len() + 5).min(isize::max_value() as usize));
+        vec.resize(4, 0);
+        LittleEndian::write_u32(&mut vec, input.len() as u32);
+        match e.read_to_end(&mut vec) {
+            Ok(_) => {
+                // according to MySQL doc: append "." if ends with space
+                if vec[vec.len() - 1] == 32 {
+                    vec.push(b'.');
+                }
+                Ok(Some(Cow::Owned(vec)))
+            }
+            _ => Ok(None),
+        }
+    }
+
+    #[inline]
+    pub fn uncompress<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        use byteorder::{ByteOrder, LittleEndian};
+        let input = try_opt!(self.children[0].eval_string(ctx, row));
+        if input.is_empty() {
+            return Ok(Some(Cow::Borrowed(b"")));
+        }
+        if input.len() <= 4 {
+            ctx.warnings.append_warning(Error::zlib_data_corrupted());
+            return Ok(None);
+        }
+
+        let len = LittleEndian::read_u32(&input[0..4]) as usize;
+        let mut decoder = ZlibDecoder::new(&input[4..]);
+        let mut vec = Vec::with_capacity(len);
+        // if the length of uncompressed string is less than the length we read from the first
```

Comment:
This comment seems wrong:
"the length of uncompressed string" = `decoded_len`
"the length we read from the first four bytes" = `len`
It is totally fine if `decoded_len < len`.